### DISCUSSION TITLE
Modular modern OpenGL rendering pipeline

### DISCUSSION THREAD

**Discussion by theottm** [kaveh808/kons-9#250] [Ideas]
The idea is to be able to run multiple passes offscreen. The aim is to be able to describe a rendering pipeline in CL and to modify it on the fly.

Rough plan:
1. write a scene renderer using OpenGL arrays
2. allow the main draw function to be changed on the fly
3. create an abstraction to describe a multi-pass pipeline
4.  describe shaders in a CL DSL

**Comment by kaveh808** [comment]
Sounds like an excellent plan. I have made an effort to isolate the OpenGL rendering calls in `kons-9/src/graphics/opengl/opengl.lisp`

**Reply by theottm** [reply]
Neat!
I played a bit around earlier and could run some arbitrary gl draw commands by evaluaring a modified version of the code in `show-window` where I replaced the call to `draw-scene-view` by a progn with my commands.
This is very rough and I would rather use a function there, say `my-draw` and redefine this function from my REPL to be able to experiment.
But my knowledge of CL stopped me there because when I defined this function it couldn't be found in the scope of kons-9. Probably something to do with in-package... I need to go read some docs.

How do you actually go about when developping ? For now I have a emacs with sly and a buffer for the source and a playground buffer. That's usually how I develop in Elisp, but there is only one giant global scope there so to speek.

**Reply by kaveh808** [reply]
Look at `test/demo-kernel.lisp` for example usage.

At the top of your file put `(in-package #:kons-9)`.

In the repl do:

`(in-package :kons-9)`

`(run)`

Then use `C-c C-c` to define your new code.
